sys/kern/pty.c

```diff
@@ -0,0 +1,273 @@
+#include <sys/filedesc.h>
+#include <sys/mutex.h>
+#include <sys/ringbuf.h>
+#include <sys/ttycom.h>
+#include <stdatomic.h>
+#include <stdio.h>
+#include <sys/condvar.h>
+#include <sys/mimiker.h>
+#include <sys/file.h>
+#include <sys/proc.h>
+#include <sys/fcntl.h>
+#include <sys/errno.h>
+#include <sys/stat.h>
+#include <sys/devfs.h>
+#include <sys/linker_set.h>
+#include <sys/tty.h>
+
+#define MAX_PTYS 16
+
+/* Must be greater than or equal to the length of the path to any slave device.
+ * Right now slave devices are at /dev/pts/<number>, so this should suffice. */
+#define PTY_PATH_MAX_LEN 15
+
+static devfs_node_t *pts_dir;
+
+typedef struct {
+  atomic_int pt_number; /* PTY number, if allocated. -1 means free. */
+  condvar_t pt_incv;    /* CV for readers */
+  condvar_t pt_outcv;   /* CV for writers */
+} pty_t;
+
+static pty_t pty_array[MAX_PTYS];
+
+static pty_t *pty_alloc(void) {
+  for (int i = 0; i < MAX_PTYS; i++) {
+    pty_t *pty = &pty_array[i];
+    if (atomic_compare_exchange_strong(&pty->pt_number, &(int){-1}, i + 1))
+      /* Got a free PTY. */
+      return pty;
+  }
+  return NULL;
+}
+
+static void pty_free(pty_t *pty) {
+  pty->pt_number = -1;
+}
+
+static int pty_read(file_t *f, uio_t *uio) {
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+  int error;
+  size_t start_resid = uio->uio_resid;
+
+  if (uio->uio_resid == 0)
+    return 0;
+
+  SCOPED_MTX_LOCK(&tty->t_lock);
+
+  /* Wait until there is at least one byte of data. */
+  while (ringbuf_empty(&tty->t_outq)) {
+    /* Don't wait for data if slave device isn't opened. */
+    if (!tty_opened(tty))
+      return 0;
+    if (cv_wait_intr(&pty->pt_incv, &tty->t_lock))
+      return ERESTARTSYS;
+  }
+
+  /* Data is available: transfer as much as we can. */
+  error = ringbuf_read(&tty->t_outq, uio);
+  tty_getc_done(tty);
+
+  /* Don't report errors on partial reads. */
+  if (start_resid > uio->uio_resid)
+    error = 0;
+
+  return error;
+}
+
+static int pty_putc_sleep(tty_t *tty, pty_t *pty, uint8_t c) {
+  while (!tty_input(tty, c)) {
+    if (cv_wait_intr(&pty->pt_outcv, &tty->t_lock))
+      return ERESTARTSYS;
+    if (!tty_opened(tty))
+      return ENOTTY;
```

Comment:
Could you explain to me why we check it in this order?

Reply:
Actually, we should check this before calling `tty_input()`, and after returning from `cv_wait_intr()`, since the tty's open count can change while we're sleeping. I have fixed it. Thanks!

Reply:
Actually, I was wrong: we don't want to check this before calling `tty_input()` the first time. If it succeeds the first time, the whole function should return 0. However, if `tty_input()` fails and we need to sleep, then we should sleep only if the tty has users, i.e. `tty_opened(tty)`. So the check should be before calling `cv_wait_intr()`: the original order was wrong. I think I have fixed it now. Please resolve this conversation once you have read this.